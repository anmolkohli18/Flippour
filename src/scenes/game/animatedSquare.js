// @flow

import React, { PureComponent } from "react";
import { Animated } from "react-native";
import {
  sequence,
  parallel,
  AnimatedView,
  animate
} from "./animatedSquare.utilities";
import type {
  AnimatePropsType,
  nextPropsType
} from "./animatedSquare.types.js";

const initial = {
  scale: 0.5,
  opacity: 0,
  rotate: -0.25
};

export class AnimatedSquare extends PureComponent {
  animating = false;

  state = {
    opacity: new Animated.Value(initial.opacity),
    scale: new Animated.Value(initial.scale),
    rotate: new Animated.Value(initial.rotate)
  };

  /**
   * Animations
   */

  levelBegin = () => {
    return parallel([
      animate(this.state.opacity, 1, 250),
      animate(this.state.scale, 1, 250),
      animate(this.state.rotate, 0, 250)
    ]);
  };

  correct = () => {
    return sequence([
      parallel([
        sequence([
          animate(this.state.scale, 0.9, 125),
          animate(this.state.scale, 1, 125)
        ]),
        animate(this.state.rotate, 1.25, 850),
        animate(this.state.opacity, 0, 850)
      ]),
      parallel([
        animate(this.state.scale, initial.scale, 0),
        animate(this.state.rotate, initial.rotate, 0),
        animate(this.state.opacity, initial.opacity, 0)
      ])
    ]);
  };

  incorrect = () => {
    return sequence([
      sequence([
        animate(this.state.scale, 0.9, 125),
        animate(this.state.scale, 1, 125),
        animate(this.state.opacity, 0, 750)
      ]),
      parallel([
        animate(this.state.scale, initial.scale, 0),
        animate(this.state.rotate, initial.rotate, 0),
        animate(this.state.opacity, initial.opacity, 0)
      ])
    ]);
  };

  fadeOut = () => {
    return sequence([
      parallel([
        animate(this.state.scale, initial.scale, 500),
        animate(this.state.opacity, 0, 500)
      ]),
      parallel([
        animate(this.state.scale, initial.scale, 0),
        animate(this.state.rotate, initial.rotate, 0),
        animate(this.state.opacity, initial.opacity, 0)
      ])
    ]);
  };

  /**
   * Lifecycle
   */

  componentDidMount = () => {
    this.levelBegin().start();
  };

  shouldComponentUpdate = (nextProps: nextPropsType) => {
    const { clicked, complete, fail, color } = nextProps;
    return (
      fail ||
      complete ||
      clicked !== this.props.clicked ||
      color !== this.props.color
    );
  };

  componentWillReceiveProps = (nextProps: nextPropsType) => {
    const { clicked, status, complete, fail, color } = nextProps;

    if (
      (color !== this.props.color && !complete) ||
      (!clicked && !complete && !fail)
    ) {
      // new level & new color
      this.animating = false;
      this.levelBegin().start();
    } else if (fail && !this.props.clicked && clicked) {
      // fail & last square chosen
      this.animating = true;
      this.incorrect().start();
      // why are we doing this?
      // to ensure we auto back on game over
      this.setState(state => state);
    } else if (fail) {
      // fail
      this.animating = true;
      this.fadeOut().start();
    } else if (complete && !this.props.clicked && clicked) {
      // complete & last square chosen
      this.animating = true;
      this.correct().start();
    } else if (complete && !this.animating) {
      // complete
      this.animating = true;
      this.fadeOut().start();
    } else if (clicked && status && !this.animating) {
      // correct
      this.animating = true;
      this.correct().start();
    } else if (clicked && !status && !this.animating) {
      // incorrect
      this.animating = true;
      this.incorrect().start();
    }
  };

  render = () => {
    const rotateYValue = this.state.rotate.interpolate({
      inputRange: [0, 1],
      outputRange: ["0deg", "360deg"]
    });

    const opacityValue = this.state.opacity.interpolate({
      inputRange: [0, 1],
      outputRange: [0, 1]
    });

    return (
      <AnimatedView
        style={{
          flex: 1,
          margin: 4,
          borderRadius: 10,
          opacity: opacityValue,
          backgroundColor: this.props.color,
          transform: [
            { scale: this.state.scale },
            { rotateY: rotateYValue },
            { perspective: 1000 }
          ]
        }}
      />
    );
  };
}
