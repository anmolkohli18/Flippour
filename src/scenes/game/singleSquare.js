import React, { PureComponent } from "react";
import { AnimatedSquare as Square } from "./animatedSquare";
import { SquareButton as Button, SquareContainer as Container } from "./styles";

export class SingleSquare extends PureComponent {
  fail = false;
  complete = false;
  clicked = false;
  status = false;
  color = "transparent";

  shouldComponentUpdate({ square, fail, complete }) {
    let shouldUpdate = false;

    // new board
    if (this.color !== square.color) {
      this.color = square.color;
      shouldUpdate = true;
    }

    // this square has been clicked
    if (this.status !== square.status || this.clicked !== square.clicked) {
      this.status = square.status;
      this.clicked = square.clicked;
      shouldUpdate = true;
    }

    // game over
    if (this.fail !== fail) {
      this.fail = fail;
      shouldUpdate = true;
    } else if (this.complete !== complete) {
      this.complete = complete;
      shouldUpdate = true;
    }

    return shouldUpdate;
  }

  onPress = () => {
    setImmediate(() => {
      this.props.onSquareClick(this.props.x, this.props.y);
    });
  };

  render = () => {
    return (
      <Button
        activeOpacity={1}
        underlayColor="transparent"
        onPress={this.onPress}
      >
        <Container>
          <Square
            fail={this.fail}
            complete={this.complete}
            status={this.status}
            clicked={this.clicked}
            color={this.props.gameColors[this.color]}
            x={this.props.x}
            y={this.props.y}
          />
        </Container>
      </Button>
    );
  };
}
