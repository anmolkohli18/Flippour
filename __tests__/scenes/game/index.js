import "react-native";
import React from "react";
import renderer from "react-test-renderer";

import { AnimatedSquare } from "../../../src/scenes/game/animatedSquare.js";
import { SingleSquare } from "../../../src/scenes/game/singleSquare.js";
import { GameScene } from "../../../src/scenes/game/index.js";

test("AnimatedSquare renders correctly", () => {
  const tree = renderer.create(<AnimatedSquare color="purple" />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("SingleSquare renders correctly", () => {
  const tree = renderer.create(<SingleSquare x={99} y={24} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("GameScene renders correctly", () => {
  const tree = renderer
    .create(
      <SingleSquare
        squares={[[{ clicked: false, status: false, color: "purple" }]]}
        complete={false}
        fail={false}
        onSquareClick={() => {}}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
