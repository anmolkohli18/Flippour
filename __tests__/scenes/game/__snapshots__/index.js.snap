// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimatedSquare renders correctly 1`] = `
<View
  renderToHardwareTextureAndroid={true}
  shouldRasterizeIOS={true}
  style={
    Object {
      "backgroundColor": "purple",
      "borderRadius": 10,
      "flex": 1,
      "margin": 4,
      "opacity": 0,
      "transform": Array [
        Object {
          "scale": 0.5,
        },
        Object {
          "rotateY": "-90deg",
        },
        Object {
          "perspective": 1000,
        },
      ],
    }
  }
/>
`;

exports[`GameScene renders correctly 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
      },
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
        },
        undefined,
      ],
    ]
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
        },
        undefined,
      ]
    }
  >
    <View
      renderToHardwareTextureAndroid={true}
      shouldRasterizeIOS={true}
      style={
        Object {
          "backgroundColor": "transparent",
          "borderRadius": 10,
          "flex": 1,
          "margin": 4,
          "opacity": 0,
          "transform": Array [
            Object {
              "scale": 0.5,
            },
            Object {
              "rotateY": "-90deg",
            },
            Object {
              "perspective": 1000,
            },
          ],
        }
      }
    />
  </View>
</View>
`;

exports[`SingleSquare renders correctly 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
      },
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
        },
        undefined,
      ],
    ]
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
        },
        undefined,
      ]
    }
  >
    <View
      renderToHardwareTextureAndroid={true}
      shouldRasterizeIOS={true}
      style={
        Object {
          "backgroundColor": "transparent",
          "borderRadius": 10,
          "flex": 1,
          "margin": 4,
          "opacity": 0,
          "transform": Array [
            Object {
              "scale": 0.5,
            },
            Object {
              "rotateY": "-90deg",
            },
            Object {
              "perspective": 1000,
            },
          ],
        }
      }
    />
  </View>
</View>
`;
